File: mobile/android/src/main/kotlin/com/otravelz/android/data/repository/PlannerRepository.kt
package com.otravelz.android.data.repository

import com.otravelz.android.data.network.ApiClient
import com.otravelz.android.data.network.NetworkError
import com.otravelz.android.data.network.NetworkResult
import com.otravelz.android.data.network.OTravelzApiService
import com.otravelz.android.data.network.dto.AIConverseRequestDto
import com.otravelz.android.data.network.dto.AIConverseResponseDto
import com.otravelz.android.data.network.dto.ChatMessageDto
import com.otravelz.android.domain.model.PlanConstraints
import com.otravelz.android.domain.model.PlanResult

/**
 * Repository orchestrating deterministic itinerary generation and grounded AI companion queries.
 */
class PlannerRepository(
    private val apiService: OTravelzApiService = ApiClient.createService()
) {

    /**
     * Executes deterministic facts-only itinerary generation via POST /itinerary/plan.
     */
    suspend fun planItinerary(constraints: PlanConstraints): NetworkResult<PlanResult> {
        val requestDto = constraints.toRequestDto()
        return when (val result = ApiClient.safeApiCall { apiService.planItinerary(requestDto) }) {
            is NetworkResult.Success -> {
                val planResult = PlanResult.fromDto(
                    dto = result.data,
                    originalConstraints = constraints
                )
                NetworkResult.Success(planResult)
            }
            is NetworkResult.Failure -> {
                result
            }
        }
    }

    /**
     * Interprets traveler intent in plain language and extracts structured planning constraints.
     */
    suspend fun extractConstraintsWithAI(
        userNote: String,
        currentConstraints: PlanConstraints
    ): NetworkResult<Pair<PlanConstraints, AIConverseResponseDto>> {
        val requestDto = AIConverseRequestDto(
            messages = listOf(ChatMessageDto(role = "user", content = userNote)),
            constraints = currentConstraints.toRequestDto()
        )

        return when (val result = ApiClient.safeApiCall { apiService.converseWithAI(requestDto) }) {
            is NetworkResult.Success -> {
                val response = result.data
                val extracted = response.constraints?.let {
                    PlanConstraints.fromRequestDto(it)
                } ?: currentConstraints

                NetworkResult.Success(Pair(extracted, response))
            }
            is NetworkResult.Failure -> {
                result
            }
        }
    }
}

File: mobile/ios/OTravelz/Features/PlannerRepository.swift
import Foundation

/// Repository coordinating deterministic plan generation and grounded conversational extraction.
public actor PlannerRepository {
    private let apiClient: APIClient

    public init(apiClient: APIClient = APIClient()) {
        self.apiClient = apiClient
    }

    /// Executes deterministic facts-only itinerary generation.
    public func planItinerary(constraints: PlanConstraints) async throws -> PlanResult {
        let requestDTO = constraints.toRequestDTO()
        let responseDTO = try await apiClient.planItinerary(request: requestDTO)
        return PlanResult.fromDTO(responseDTO, originalConstraints: constraints)
    }

    /// Interprets natural language traveler notes and extracts structured planning constraints.
    public func extractConstraintsWithAI(
        userNote: String,
        currentConstraints: PlanConstraints
    ) async throws -> (PlanConstraints, AIConverseResponseDTO) {
        let requestDTO = AIConverseRequestDTO(
            messages: [ChatMessageDTO(role: "user", content: userNote)],
            constraints: currentConstraints.toRequestDTO()
        )

        let responseDTO = try await apiClient.converseWithAI(request: requestDTO)
        let extracted: PlanConstraints = {
            if let c = responseDTO.constraints {
                return PlanConstraints.fromRequestDTO(c)
            }
            return currentConstraints
        }()

        return (extracted, responseDTO)
    }
}
